evals/cybench/dataset.py

```diff
@@ -0,0 +1,150 @@
+from pathlib import Path
+from typing import Callable
+
+import yaml
+
+from inspect_ai.dataset import Dataset, MemoryDataset, Sample
+
+SUPPORTED_FILE_NAMES = ["challenge.yaml", "challenge.yml"]
+
+
+def cybench_dataset(
+    challenges_dir: Path | list[Path] | None,
+    filters: Callable[[Sample], bool] | list[Callable[[Sample], bool]] | None,
+) -> Dataset:
+    # If no path specified, run all the challenges.
+    if not challenges_dir:
+        challenges_dir = Path(__file__).resolve().parent
+    challenges_dir = (
+        challenges_dir if isinstance(challenges_dir, list) else [challenges_dir]
+    )
+    challenges = _get_challenges(challenges_dir)
+    dataset = MemoryDataset(samples=_create_samples(challenges))
+    if filters:
+        dataset = _filter_dataset(dataset, filters)
+    return dataset
+
+
+def _get_challenges(challenges_dirs: list[Path]) -> list[Path]:
+    def get_challenges(challenge_dir: Path) -> list[Path]:
+        if not isinstance(challenge_dir, Path):
+            raise TypeError(
+                "Expected challenge_dir to be a Path, "
+                f"but got {type(challenge_dir).__name__}."
+            )
+
+        challenge_dir = challenge_dir.resolve()
+        if not challenge_dir.exists():
+            raise FileNotFoundError(f"Could not find: {challenge_dir}")
+
+        challenge_dirs = set()
+        for file_name in SUPPORTED_FILE_NAMES:
+            for file in challenge_dir.rglob(file_name):
+                challenge_dirs.add(file.parent)
+        return list(challenge_dirs)
+
+    all_challenge_dirs = set()
+    for challenge_dir in challenges_dirs:
+        all_challenge_dirs.update(get_challenges(challenge_dir))
+
+    return list(all_challenge_dirs)
+
+
+def _create_samples(challenges: list[Path]) -> list[Sample]:
+    def create_sample_per_variant(challenge: Path) -> list[Sample]:
+        challenge_info = _load_challenge_info(challenge)
```

Comment:
At present, you're reloading the whole yaml file inside the variant loop. I'm going to hack out a refactor for this `_create_samples()` function to break it down into smaller parts.

Reply:
Ah, apologies, I think I misunderstood the loop. This is actually just called once per challenge. Ignore me!